File: apps/backend/app/models/visitor_training.ts
import { DateTime } from 'luxon'
import { BaseModel, column, belongsTo, hasMany } from '@adonisjs/lucid/orm'
import type { BelongsTo, HasMany } from '@adonisjs/lucid/types/relations'
import StudentEnroll from '#models/student_enroll'
import Instructor from '#models/instructor'
import VisitorSchedule from '#models/visitor_schedule'
import VisitorEvaluateCompany from './visitor_evaluate_company.js'
import VisitorEvaluateStudent from './visitor_evaluate_student.js'

export default class VisitorTraining extends BaseModel {
  public static table = 'visitor_trainings'

  @column({ isPrimary: true })
  declare id: number

  @column()
  declare student_enroll_id: number

  @column()
  declare visitor_instructor_id: number

  @column.dateTime({ autoCreate: true })
  declare createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  declare updatedAt: DateTime

  // relations
  @belongsTo(() => StudentEnroll, {
    foreignKey: 'student_enroll_id',
  })
  declare studentEnroll: BelongsTo<typeof StudentEnroll>

  @belongsTo(() => Instructor, {
    foreignKey: 'visitor_instructor_id',
  })
  declare visitor: BelongsTo<typeof Instructor>

  @hasMany(() => VisitorSchedule, {
    foreignKey: 'visitor_training_id',
  })
  declare schedules: HasMany<typeof VisitorSchedule>

  @hasMany(() => VisitorEvaluateCompany, {
    foreignKey: 'student_training_id',
  })
  declare evaluateCompany: HasMany<typeof VisitorEvaluateCompany>

  @hasMany(() => VisitorEvaluateStudent, {
    foreignKey: 'visitor_training_id',
  })
  declare evaluateStudent: HasMany<typeof VisitorEvaluateStudent>
}

File: apps/backend/internal/models/visitor_training.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// VisitorTraining represents the visitor_trainings table
type VisitorTraining struct {
	ID                  uint      `gorm:"primaryKey;autoIncrement" json:"id"`
	StudentEnrollID     uint      `gorm:"column:student_enroll_id;not null" json:"student_enroll_id"`
	VisitorInstructorID uint      `gorm:"column:visitor_instructor_id;not null" json:"visitor_instructor_id"`
	CreatedAt           time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt           time.Time `gorm:"autoUpdateTime" json:"updated_at"`

	// Relationships
	StudentEnroll        StudentEnroll          `gorm:"foreignKey:StudentEnrollID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"student_enroll,omitempty"`
	Visitor              Instructor             `gorm:"foreignKey:VisitorInstructorID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE" json:"visitor,omitempty"`
	Schedules            []VisitorSchedule      `gorm:"foreignKey:VisitorTrainingID" json:"schedules,omitempty"`
	EvaluateCompany      []VisitorEvaluateCompany `gorm:"foreignKey:VisitorTrainingID" json:"evaluate_company,omitempty"`
	EvaluateStudent      []VisitorEvaluateStudent `gorm:"foreignKey:VisitorTrainingID" json:"evaluate_student,omitempty"`
}

// TableName specifies the table name for VisitorTraining model
func (VisitorTraining) TableName() string {
	return "visitor_trainings"
}

// BeforeDelete hook to clean up related records when visitor training is deleted
func (vt *VisitorTraining) BeforeDelete(tx *gorm.DB) error {
	// Delete all visitor schedules for this training
	if err := tx.Where("visitor_training_id = ?", vt.ID).Delete(&VisitorSchedule{}).Error; err != nil {
		return err
	}

	// Delete all visitor evaluations for companies
	if err := tx.Where("visitor_training_id = ?", vt.ID).Delete(&VisitorEvaluateCompany{}).Error; err != nil {
		return err
	}

	// Delete all visitor evaluations for students
	if err := tx.Where("visitor_training_id = ?", vt.ID).Delete(&VisitorEvaluateStudent{}).Error; err != nil {
		return err
	}

	return nil
}